Mathlib/Algebra/Category/AlgebraCat/Basic.lean

```diff
@@ -46,22 +47,97 @@ instance : CoeSort (AlgebraCat R) (Type v) :=
 
 attribute [coe] AlgebraCat.carrier
 
+/-- The object in the category of R-algebras associated to a type equipped with the appropriate
+typeclasses. This is the preferred way to construct a term of `AlgebraCat R`. -/
+abbrev of (X : Type v) [Ring X] [Algebra R X] : AlgebraCat.{v} R :=
+  ⟨X⟩
+
+lemma coe_of (X : Type v) [Ring X] [Algebra R X] : (of R X : Type v) = X :=
+  rfl
+
+variable {R} in
+/-- The type of morphisms in `AlgebraCat R`. -/
+@[ext]
+structure Hom (A B : AlgebraCat.{v} R) where
+  private mk ::
+  /-- The underlying algebra map. -/
+  hom : A →ₐ[R] B
+
 instance : Category (AlgebraCat.{v} R) where
-  Hom A B := A →ₐ[R] B
-  id A := AlgHom.id R A
-  comp f g := g.comp f
+  Hom A B := Hom A B
+  id A := ⟨AlgHom.id R A⟩
+  comp := fun ⟨f⟩ ⟨g⟩ ↦ ⟨g.comp f⟩
```

Comment:
I claim this was better as
```suggestion
  comp := fun  f g ↦ f.hom.comp g.hom
```
as it avoids inducing `match` terms into the expression (and throughout). I think @mattrobball suggested the opposite though.

Reply:
Yes, this was the original version and the change caused slowdowns, which I had reverted and then restored upon request by Matthew.

Reply:
I think `simp` has better support for `match` reduction than for eta reduction in situations like this. I linked to a thread on Zulip with a basic example.

Reply:
Does this better (automatic) `simp` support really matter though in a situation where we supply the relevant `simp` lemmas anyway?

Reply:
Not having to always provide "obvious" simp lemmas seems a worthwhile goal.

Reply:
Agreed, here we have them though. I don't have a strong opinion on this, the only empirical evidence for me is the (mild) slowdown favoring the original version.

Reply:
> I linked to a thread on Zulip with a basic example.

Would you mind repeating it here?

Reply:
It is this one: https://leanprover.zulipchat.com/#narrow/channel/270676-lean4/topic/optimal.20instance.20design.20for.20structure.20wrappers